kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -332,6 +359,91 @@ static SchemaChanges computeSchemaChangesById(
     return schemaDiff.build();
   }
 
+  private static void validatePhysicalNameConsistency(
+      List<Tuple2<StructField, StructField>> updatedFields) {
+    for (Tuple2<StructField, StructField> updatedField : updatedFields) {
+      StructField currentField = updatedField._1;
+      StructField newField = updatedField._2;
+      if (!getPhysicalName(currentField).equals(getPhysicalName(newField))) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Existing field with id %s in current schema has "
+                    + "physical name %s which is different from %s",
+                getColumnId(currentField),
+                getPhysicalName(currentField),
+                getPhysicalName(newField)));
+      }
+    }
+  }
+
+  /* Validate if a given schema evolution is safe for a given column mapping mode*/
+  private static void validateSchemaEvolution(
+      StructType currentSchema, StructType newSchema, Metadata metadata) {
+    ColumnMappingMode columnMappingMode =
+        ColumnMapping.getColumnMappingMode(metadata.getConfiguration());
+    switch (columnMappingMode) {
+      case ID:
+      case NAME:
+        validateSchemaEvolutionById(currentSchema, newSchema, metadata);
+        return;
+      case NONE:
+        throw new UnsupportedOperationException(
+            "Schema evolution without column mapping is not supported");
+      default:
+        throw new UnsupportedOperationException(
+            "Unknown column mapping mode: " + columnMappingMode);
+    }
+  }
+
+  /* Validates a given schema evolution by using field ID
+  as the source of truth for identifying fields */
+  private static void validateSchemaEvolutionById(
+      StructType currentSchema, StructType newSchema, Metadata metadata) {
+    Map<Long, StructField> currentFieldsById = fieldsById(currentSchema);
+    Map<Long, StructField> updatedFieldsById = fieldsById(newSchema);
+    SchemaChanges schemaChanges = computeSchemaChangesById(currentFieldsById, updatedFieldsById);
+
+    // Validate physical name consistency on the updated fields
+    validatePhysicalNameConsistency(schemaChanges.updatedFields());
+
+    // ToDo Validate IcebergCompatV2 nested IDs
+
+    // ToDo: Validate schema compatibility
+  }
+
+  /** Returns a map from field ID to the field in the schema */
+  private static Map<Long, StructField> fieldsById(StructType schema) {
```

Comment:
At some point, we may want to preserve the full path (maybe Map<Long, Pair<String, StructField>>. I don't think that'd ever be used as a source of truth in validation but it'd probably be useful in failure messages.